js/ui/workspaceMonitor.js

```diff
@@ -10,276 +10,144 @@ const WorkspaceMonitor = new Lang.Class({
         }
     },
 
-    _minimizeWindow: function(shellwm, actor) {
-        if (!this._windowIsKnown(actor.meta_window)) {
-            return;
-        }
-
-        this._windowGoingInvisible();
-
-        // We'll show the overview when the destroy animation ends
-        this._addMinimizedWindow(actor.meta_window);
+    _setAppVisible: function(app, visible) {
+        if (visible)
+            this._visibleApps.add(app);
+        else
+            this._visibleApps.delete(app);
     },
 
-    _minimizeWindowCompleted: function(shellwm, actor) {
-        if (this._windowIsMinimized(actor.meta_window)) {
-            this._windowGoingInvisibleCompleted();
-            this._removeMinimizedWindow(actor.meta_window);
+    _onAppStateChange: function(appSystem, app) {
+        if (app.get_state() == Shell.AppState.RUNNING) {
+            this._trackApp(app);
         }
     },
 
-    _realMapWindow: function(metaWindow) {
-        // If we don't know about it then we don't handle it
-        if (!this._windowIsKnown(metaWindow)) {
-            return;
-        }
+    _onAppWindowsChanged: function(app) {
+        this._setAppVisible(app, this._appHasVisibleWindows(app));
 
-        // If the window was minimized remove it
-        if (this._windowIsMinimized(metaWindow)) {
-            this._removeMinimizedWindow(metaWindow);
+        // We need to track the stopped state here (instead of in _onAppStateChange),
+        // and independently from setting an app visible, because there are apps
+        // that go into "running" -> "stopped" -> "running" states because of the
+        // way they show their windows. This is the case of the Maps (marble) app.
+        if (app.get_state() == Shell.AppState.STOPPED) {
+            let has_visible_windows = this._appHasVisibleWindows(app);
+            this._setAppVisible(app, has_visible_windows);
+            if (!has_visible_windows)
+                this._untrackApp(app);
         }
 
-        this._visibleWindows += 1;
-
         this._updateOverview();
     },
 
-    _mapWindow: function(shellwm, actor) {
-        this._realMapWindow(actor.meta_window);
-    },
-
-    _unminimizeWindow: function(shellwm, actor) {
-        this._realMapWindow(actor.meta_window);
-    },
-
-    _fullscreenChanged: function() {
-        let primaryMonitor = Main.layoutManager.primaryMonitor;
-        let inFullscreen = primaryMonitor && primaryMonitor.inFullscreen;
-
-        if (this._inFullscreen != inFullscreen) {
-            this._inFullscreen = inFullscreen;
-            this._updateOverview();
-        }
-    },
+    _appHasVisibleWindows: function(app) {
+        let windows = app.get_windows();
+        for (let window of windows) {
+            // We do not count transient windows because of an issue with Audacity
+            // where a transient window was always being counted as visible even
+            // though it was minimized
+            if (window.get_transient_for())
```

Comment:
Interesting, why were we not hitting the same problem before? I wonder if this couldn't cause other issues though when done generically.